File: backend/models/Member.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

const MembershipType = sequelize.define('MembershipType', {
  id: { type: DataTypes.UUID, defaultValue: DataTypes.UUIDV4, primaryKey: true },
  name: { type: DataTypes.STRING, allowNull: false, unique: true },
  amount: { type: DataTypes.FLOAT, allowNull: false },
  duration_months: { type: DataTypes.INTEGER, allowNull: false, defaultValue: 1 },
}, { tableName: 'membership_types', timestamps: true });

const Member = sequelize.define('Member', {
  id: { type: DataTypes.UUID, defaultValue: DataTypes.UUIDV4, primaryKey: true },
  gym_id: { type: DataTypes.UUID, allowNull: false },
  member_name: { type: DataTypes.STRING, allowNull: false },
  phone: { type: DataTypes.STRING, allowNull: false },
  email: { type: DataTypes.STRING, validate: { isEmail: true } },
  avatar: { type: DataTypes.TEXT, allowNull: true },
  last_arrival: { type: DataTypes.DATE, allowNull: true },
  membership_type_id: { type: DataTypes.UUID, allowNull: true },
  join_date: { type: DataTypes.DATEONLY, defaultValue: DataTypes.NOW },
  expiry_date: { type: DataTypes.DATEONLY, allowNull: true },
  status: { type: DataTypes.ENUM('active', 'expired', 'trial'), defaultValue: 'active' },
  is_trial: { type: DataTypes.BOOLEAN, defaultValue: false },
  payment_collected: { type: DataTypes.BOOLEAN, defaultValue: false },
  last_payment_date: { type: DataTypes.DATE, allowNull: true, defaultValue: null },
  total_visits: { type: DataTypes.INTEGER, defaultValue: 0 },
  lifetime_value: { type: DataTypes.FLOAT, defaultValue: 0 },
}, { tableName: 'members', timestamps: true });

// --- New Model Imports ---
const WorkflowReminder = require('./WorkflowReminder');
const Call = require('./Call');
const AttendanceSession = require('./AttendanceSession');
const Payment = require('./Payment');

// Associations
Member.belongsTo(MembershipType, { foreignKey: 'membership_type_id' });
MembershipType.hasMany(Member, { foreignKey: 'membership_type_id' });

// Member <-> WorkflowReminder
Member.hasMany(WorkflowReminder, { foreignKey: 'member_id', onDelete: 'CASCADE' });
WorkflowReminder.belongsTo(Member, { foreignKey: 'member_id' });

// Member <-> Call
Member.hasMany(Call, { foreignKey: 'member_id', onDelete: 'CASCADE' });
Call.belongsTo(Member, { foreignKey: 'member_id' });

Member.hasMany(AttendanceSession, { foreignKey: 'member_id', onDelete: 'CASCADE' });
AttendanceSession.belongsTo(Member, { foreignKey: 'member_id' });

// Member <-> Payment
Member.hasMany(Payment, { foreignKey: 'member_id', onDelete: 'CASCADE' });
Payment.belongsTo(Member, { foreignKey: 'member_id' });

module.exports = { Member, MembershipType, WorkflowReminder, Call, AttendanceSession, Payment };

File: gym-membership-management/my_app/lib/models/member.dart
class MembershipType {
  final String id;
  final String name;
  final double amount;
  final int durationMonths;

  MembershipType({
    required this.id,
    required this.name,
    required this.amount,
    required this.durationMonths,
  });

  factory MembershipType.fromJson(Map<String, dynamic> json) {
    final id = json['id']?.toString() ?? '';
    final name = json['name']?.toString() ?? '';
    final amountRaw = json['amount'];
    final amount = amountRaw is num
        ? amountRaw.toDouble()
        : double.tryParse(amountRaw?.toString() ?? '') ?? 0.0;
    final durRaw = json['duration_months'];
    final durationMonths = durRaw is int
        ? durRaw
        : (durRaw is num ? durRaw.round() : int.tryParse(durRaw?.toString() ?? '') ?? 1);

    return MembershipType(
      id: id,
      name: name,
      amount: amount,
      durationMonths: durationMonths,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'amount': amount,
      'duration_months': durationMonths,
    };
  }
}

class Member {
  final String? id;
  final String? gymId;
  final String memberName;
  final String phone;
  final String? email;
  final String? membershipTypeId;
  final String? joinDate;
  final String? expiryDate;
  final String status;
  final bool isTrial;
  final bool paymentCollected;
  final int totalVisits;
  final double lifetimeValue;
  final MembershipType? membershipType;

  Member({
    this.id,
    this.gymId,
    required this.memberName,
    required this.phone,
    this.email,
    this.membershipTypeId,
    this.joinDate,
    this.expiryDate,
    this.status = 'active',
    this.isTrial = false,
    this.paymentCollected = false,
    this.totalVisits = 0,
    this.lifetimeValue = 0,
    this.membershipType,
  });

  factory Member.fromJson(Map<String, dynamic> json) {
    return Member(
      id: json['id'],
      gymId: json['gym_id'] ?? '',
      memberName: json['member_name'] ?? '',
      phone: json['phone'] ?? '',
      email: json['email'],
      membershipTypeId: json['membership_type_id'],
      joinDate: json['join_date'],
      expiryDate: json['expiry_date'],
      status: json['status'] ?? 'active',
      isTrial: json['is_trial'] ?? false,
      paymentCollected: json['payment_collected'] ?? false,
      totalVisits: json['total_visits'] ?? 0,
      lifetimeValue: (json['lifetime_value'] ?? 0).toDouble(),
      membershipType: json['MembershipType'] != null 
          ? MembershipType.fromJson(json['MembershipType']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'gym_id': gymId,
      'member_name': memberName,
      'phone': phone,
      'email': email,
      'membership_type_id': membershipTypeId,
      'is_trial': isTrial,
      'payment_collected': paymentCollected,
    };
  }
}
